uk.co.compucorp.civicrm.tasksassignments/api/v3/TASettings.php

```diff
@@ -1,70 +1,161 @@
-        ));
-        $item =  CRM_Utils_Array::first($setting['values']);
-        $result[$field] = array(
-            'name' => $item['name'],
-            //'label' => $item['label'],
-            'value' => $item['value'] != "null" ? $item['value'] : ""
-        );
+  $fields = _civicrm_api3_t_a_settings_valid_fields();
+
+  if (!empty($params['fields'])) {
+    $fields = (array) $params['fields'];
+  }
+
+  $settings = _civicrm_api3_t_a_settings_get_settings();
+
+  $result = [];
+
+  // safeguard in case this API gets called while the
+  // ta_settings setting is still not available
+  if (empty($settings['ta_settings'])) {
+    return $result;
+  }
+
+  foreach ($fields as $field) {
+    if (array_key_exists($field, $settings['ta_settings'])) {
+      $result[$field] = [
+        'name' => $field,
+        'value' => $settings['ta_settings'][$field] != 'null' ? $settings['ta_settings'][$field] : ''
+      ];
     }
-    
-    return civicrm_api3_create_success($result, $params, 'TASettings', 'get');
+  }
+
+  return civicrm_api3_create_success($result, $params, 'TASettings', 'get');
+}
+
+/**
+ * TASettings.set API specification
+ *
+ * @param array $spec
+ */
+function _civicrm_api3_t_a_settings_set_spec(&$spec) {
+  $spec['fields'] = [
+    'name' => 'fields',
+    'title' => 'Fields',
+    'description' => 'Which field(s) you want set',
+    'type' => CRM_Utils_Type::T_STRING,
+    'api.required' => 1,
+  ];
 }
 
+/**
+ * TASettings.set API
+ *
+ * Used to set a new value in the T&A Settings. Usage:
+ *
+ * civicrm_api3('TASettings', 'set', [
+ *   'field1' => 'bah',
+ *   'field2' => 'foo',
+ *   ...
+ * ])
+ *
+ * @param array $params
+ *
+ * @return array
+ * @throws \CiviCRM_API3_Exception
+ *   An Error will be thrown if any of the given fields doesn't exist
+ */
 function civicrm_api3_t_a_settings_set($params) {
-    
-    foreach ((array)$params['fields'] as $key => $value) {
-        if ($key === 'is_task_dashboard_default') {
-            if ($value == '1') {
-                CRM_Tasksassignments_DashboardSwitcher::switchToTasksAndAssignments();
-            } else {
-                CRM_Tasksassignments_DashboardSwitcher::switchToDefault();
-            }
-        }
-        if ($key === 'days_to_create_a_document_clone') {
-          $value = (int)$value;
-        }
-
-        $setting = civicrm_api3('OptionValue', 'get', array(
-            'option_group_id' => 'ta_settings',
-            'name' => $key,
-        ));
-
-        if (empty($setting['id'])) {
-            continue;
-        }
-
-        civicrm_api3('OptionValue', 'create', array(
-            'id' => $setting['id'],
-            'value' => $value,
-        ));
+  $validFields = _civicrm_api3_t_a_settings_valid_fields();
+
+  $requestFields = [];
+  if (!empty($params['fields']) && is_array($params['fields'])) {
+    $requestFields = $params['fields'];
+  }
+
+  foreach ($requestFields as $field => $value) {
+    if (!in_array($field, $validFields)) {
```

Comment:
Do you think it might be better to do the validation first before the update? In this case it would be possible to pass an array of multiple valid settings and all those up to the first invalid one would be updated

Reply:
I optimized this whole thing a little bit. Now, instead of making multiple calls to `Setting.create` it does that only once. It now works as you've suggested and the settings will not be updated if there's something wrong.